### DISCUSSION TITLE
"Required" on update

### DISCUSSION THREAD

**Discussion by dewita97** [directus/directus#10276] [Archive]
Hello, 

Apparently, when you create a mandatory field condition based on the status of an item on some fields of a collection, it is only valid when the item is created in the database. The check will not be performed after future changes. (see #10273)

How can I make sure that the fields I decide are mandatory only when I want to publish my item?
I find that the draft status doesn't make sense if you have to fill in all the fields of the item anyway before saving :/ 

Could you please enlighten me?

**Comment by rijkvanzanten** [comment]
I think the real underlying question here is:

What should _Required_ do on _update_. Right now, the flag effectively checks whether or not a value is _submitted_. This works exactly as you'd expect for _create_, but loses usefulness on _update_.

The main question becomes: When does something pass the required state. Is it when you submit _something_ versus nothing? What is _something_? Would `null` pass as a value, as you required something to be submitted? What about an empty string? The numeric value `0`? Boolean `false`?

The way you've configured the constraints in #10273 would actually be the way I would expect to set it up as well. We need to look into how to update and improve on the required flag tracking on update. I think we can't get around to adding another option to let the user decide when something passes the "Required" state with options:

- "When any value is submitted"
- "When value is non-falsy"
- "When value passes validation rules" (as configured in permissions)

(the second one would effectively mirror JavaScript falsy checks)

(cc @benhaynes )

**Reply by benhaynes** [reply]
It's another one of those things that are too subjective/opinionated for us to decide, so I think we need to err on the side of flexibility. I like the idea of having options that let the admin decide what "required" means to them and their project. This could be on a field-level, and just need a decent default so most people can just toggle it on/off.

Perhaps it's a toggle, and when activated an additional dropdown comes up for rules/context. Otherwise, it can be a dropdown that includes "Not Required" by default, and other options beyond that (this might be best).

**Reply by frheinze** [reply]
What I would expect is that "Required value to be set" at field level behaves similar to "Required value to be set" at condition level. And this is currently not the case, the former is also checked on update. Or am I misunderstanding the concept?

**Comment by kaziller** [comment]
We are still using 9.0.0-rc.69 and we have solved this by creating a hook, which listens to 'items.update.before' (now filter('items.update', () => {...}). In that hook we are checking whether the specific collection is being updated and if the status is set to "PUBLISHED". We hard coded our required fields in that hook and check, whether the fields are empty or not. If Yes, we throw a FailedValidationException and the form error message for the affected fields will be shown.

*Note: This is the old way of registering a hook

```js
const requiredFields = [
	'field1',
	'field2',
	'field3',
];

module.exports = function registerHook({ services, exceptions }) {
	return {
		'items.update.before': async function (payload, context) {
			const { ItemsService } = services;
			const { FailedValidationException } = exceptions;
			const { collection, item, schema, accountability } = context;

			if (collection !== 'example') {
				return payload;
			}

			const isBatch = Array.isArray(item);

			const exampleService = new ItemsService('example', { schema, accountability });
			const examples = isBatch
				? await Promise.all(item.map((i) => exampleService.readByKey(i)))
				: [await exampleService.readByKey(item)];

			const needsValidation =
				payload.status === 'PUBLISHED' || (!isBatch && examples[0].status === 'PUBLISHED' && !payload.status);

			const validationErrors = examples.reduce((acc, example) => {
				const missingFields = requiredFields.reduce((acc, field) => {
					if (!isNullOrUndefined(example[field])) {
						if (isNull(payload[field])) {
							acc.push(field);
						}
					} else {
						if (isNullOrUndefined(payload[field])) {
							acc.push(field);
						}
					}
					return acc;
				}, []);
				if (missingFields.length) {
					acc.push({ id: example.example_id, errors: missingFields });
				}
				return acc;
			}, []);

			if (needsValidation && validationErrors.length) {
				console.log('PATCH request interrupted. Example(s) failed validation: ', validationErrors);

				throw new FailedValidationException({
					message: `Attempted to publish example with missing required fields: ${validationErrors[0].errors.join(
						', '
					)}`,
					path: validationErrors[0].errors,
					type: 'required',
				});
			}
			return payload;
		},
	};
};
```

In the process of updating to 9.1.2 we have noticed, that we can not throw 'FailedValidationException' anymore, since it is not included in the services anymore. We have tried to throw another Exception, which works of course but you cannot highlight the affected fields in the form anymore:

*Note: Current way of registering a hook and check for required fields

```js
const requiredFields = [
	'field1',
	'field2',
	'field3',
];

module.exports = function registerHook({ filter }, { services, exceptions }) {
	filter('items.update', async (payload, { keys, collection }, { schema, accountability }) => {
	  const { ItemsService } = services;
	  const { InvalidPayloadException } = exceptions;
  
	  if (collection !== 'examples') {
		return payload;
	  }
  
	  const isBatch = Array.isArray(item);
  
	  const exampleService = new ItemsService('examples', { schema, accountability });
  
	  const examples = await Promise.all(keys.map((i) => exampleService.readOne(i)));
  
	  const needsValidation = payload.status === 'PUBLISHED' || (!isBatch && examples[0].status === 'PUBLISHED' && !payload.status);
  
	  const validationErrors = examples.reduce((acc, example) => {
		const missingFields = requiredFields.reduce((acc, field) => {
		  if (!isNullOrUndefined(example[field])) {
			if (isNull(payload[field])) {
			  acc.push(field);
			}
		  } else {
			if (isNullOrUndefined(payload[field])) {
			  acc.push(field);
			}
		  }
		  return acc;
		}, []);
		if (missingFields.length) {
		  acc.push({ id: example.example_id, errors: missingFields });
		}
		return acc;
	  }, []);
  
	  if (needsValidation && validationErrors.length) {
		console.log('PATCH request interrupted. Example(s) failed validation: ', validationErrors);
  
		throw new InvalidPayloadException('Some fields are required');
	  }
	  return payload;
	});
  };
```

In this case we are throwing an "InvalidPayloadException", which of course works but we are missing the feature of showing the form error messages. 

This is the way we did in the past, I am not suggesting to include "FailedValidationException", since I think you took it out for a reason. My preferred solution would be to handle this via the conditions. But I would expect these conditions to apply in any form, whether it is a form to create an item or to update an item.

**Reply by rijkvanzanten** [reply]
> But I would expect these conditions to apply in any form, whether it is a form to create an item or to update an item.

Exactly! That's the same as @dewita97 mentioned 🙂 I'm sure we can figure out an improvement to the Required state to fit this use case. Any thoughts regarding when the required state should pass that I outlined in https://github.com/directus/directus/discussions/10276#discussioncomment-1745224?

As for the FailedValidationException, I don't recall it being taken out on purpose. I'll look into that 👍🏻

**Reply by rijkvanzanten** [reply]
(also quick GH pro-tip, if you add `js` at the end of the triple ` you get syntax highlighting 🙂 )

**Comment by dewita97** [comment]
I am a new user of Directus, so I don't have all the keys to understand the spirit in which it was developed. If I followed correctly, one of the main advantages of this framework is to allow the quick creation of collections, all without code! The disadvantage of a "no code" solution is that you have to think exhaustively about all possible cases, and decide what is relevant to include or not. 

Concerning the status, I see 2 problems at the moment: 

The first one is that there is no intrinsic difference between "draft" and "published" status, except that a draft does not appear in the item retrieval (example: api call).

The second problem is that there is no way to natively configure the behavior of this type of state in the CMS administration (in no code)

Of course, Directus is well thought out, and allows to extend/modify some functionalities in order to reach our requirements via extensions. Nevertheless, don't you find something missing in the basic solution proposed by the framework?

The solution proposed by @kaziller (whose sharing I thank) uses this extension system but it remains in my opinion a workaround to counter these 2 problems. I feel that the framework should propose a native solution to this problem, whether it is configurable by the administrator or not. 

Let's take for example the case of WordPress, whose status system I invite you to consult on this page: https://wordpress.org/support/article/post-status/

Natively, 8 statuses are proposed, with a well defined utility for each entity. In the case of the draft, the post corresponds to an item that does not have all its fields completed or at least is not ready to be published. Of course, new statuses can be added (as is the case for Directus), but the essentials are present on WordPress and are sufficient to manage the vast majority of cases. For a management system as popular as this one and intended for developers and not tech savvy, this solution is acceptable. 

Now let's see what the Laravel framework offers to solve this problem. Well, nothing at all! We have to add this functionality ourselves via a package or our code. Laravel is intended for a community of developers, because it is only configurable through traditional programming. Basically, there is nothing, so we have to manage and write the migrations, rules, routes,... (although all the tools at our disposal allow us to do all this very quickly). In this paradigm, it is reasonable to accept this solution, because it is up to the developer to create his own custom solution. 

To come back to Directus, we are a bit between the 2 extremes that are WordPress and Laravel. It is not useful to create all possible statuses with a particular behavior for the creation of an item, because it would be too subjective/opinionated. You have to boil it down to the essentials! On the other hand, not proposing a solution to the problem, or at least making status available without the possibility of configuring them, seems to break the configurable spirit that "no code" offers.

From this point of view, should the first problem be solved in order to solve the second or the opposite? Should we start a completely different debate? @rijkvanzanten @benhaynes

**Reply by benhaynes** [reply]
Thank you for the thoughtful and comprehensive comment, @dewita97!

At a high-level, we first aim to get the codebase customizable and extensible. This provides the escape hatch and allows the most flexibility. It also helps us build the simplified GUI because we better understand how the feature will work and how users intend to use it. Also, the underlying code is important to figure out (agnostically) before diving into the UI (for the same reason we needed to build the API first, to power the App second).

Once we have that, we use that knowledge to build an interface with good defaults and helpful no-code options.

If we can find a set of intelligent and unopinionated `status` options, we can look into adding these as out-of-the-box options for faster admin setup. Things to keep in mind:

* Options should pass the 80/20 rule. To that end, to keep the APp simple and intuitive, we don't want to busy the GUI with too many options/configurations, trading functionality for a confusing experience.
* We should think beyond the single/primary use-case (eg Content in a CMS) and consider things though a broader "data" lens
* We should ensure there is a proper escape hatch for anything we don't include

I'd love to hear your thoughts on a balanced set of status options... as well as how each manifests throughout the platform. And whenever possible, we should leverage existing functionality. Meaning, we should ty to use our permissions system for most of these, perhaps offering "helpers" to set permissions from other configuration points. This avoids a lot of custom code that needs to be maintained and centralizes a lot of this logic in one place.

**Comment by janwillemvd** [comment]
It might be an idea to rename this option to 'Required value to be set **on creation**'?

![Screenshot 2022-05-17 at 11 34 26](https://user-images.githubusercontent.com/10515908/168779949-c01e48d6-40f7-427f-a4d7-0f1e4e5e234a.png)

If someone then wants to make this value required on updates too, one can use the Validations.
